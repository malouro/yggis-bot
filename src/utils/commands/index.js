import canUseCommand from './canUseCommand'
import getCommandFromMessage from './getCommandFromMessage'
import parseArguments from './parseArguments'

export {
	canUseCommand,
	getCommandFromMessage,
	parseArguments,
}
