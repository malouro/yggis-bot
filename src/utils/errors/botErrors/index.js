export { default as BotError } from './BotError';
export { default as MissingTokenError } from './MissingTokenError';
