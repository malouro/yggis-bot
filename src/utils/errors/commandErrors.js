class CommandError extends Error {
	constructor(message) {
		super(message)

		this.name = 'CommandError'
	}
}

class MissingPropertyError extends CommandError {
	constructor(property) {
		super(`No property \`${property}\``)

		this.name = 'MissingProperty'
		this.property = property
	}
}

export {
	CommandError,
	MissingPropertyError,
}
