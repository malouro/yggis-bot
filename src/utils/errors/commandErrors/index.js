export { default as CommandError } from './CommandError'
export { default as InvalidPropertyError } from './InvalidPropertyError'
export { default as MissingPropertyError } from './MissingPropertyError'
