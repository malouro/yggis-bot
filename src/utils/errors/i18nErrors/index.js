export { default as I18nError } from './I18nError';
export { default as MissingI18nConfigError } from './MissingI18nConfig';
