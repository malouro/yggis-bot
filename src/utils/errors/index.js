export {
	CommandError,
	MissingPropertyError,
	InvalidPropertyError,
} from './commandErrors';
export { BotError, MissingTokenError } from './botErrors';
