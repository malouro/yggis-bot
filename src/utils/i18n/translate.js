import { MissingI18nConfigError } from '../errors';

export default ({ defaultSpace = 'COMMON', translations, language }) => {
	if (!translations || !language) {
		throw new MissingI18nConfigError();
	}

	return (space = defaultSpace, subspace, key) => {
		let output;

		try {
			if (subspace) {
				output = translations[language][space][subspace][key];
			} else {
				output = translations[language][space][key];
			}
		} catch (err) {
			/* eslint-disable-next-line no-console */
			console.error(err);
			output = null;
		}

		return output;
	};
};
