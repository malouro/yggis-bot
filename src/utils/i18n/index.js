import translate from './translate';

/* eslint-disable-next-line import/prefer-default-export */
export { translate };
