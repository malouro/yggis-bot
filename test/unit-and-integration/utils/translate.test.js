import mockConsole from 'jest-mock-console';

import translate from '../../../src/utils/i18n/translate';
import { MissingI18nConfigError } from '../../../src/utils/errors';

const testTranslations = {
	test: {
		COMMON: {
			someSubspace: {
				keyWithinSubspace: 'Key within a subspace',
				functionWithinSubspace: jest.fn(
					val => `Function within a subspace called with ${val}`
				),
			},
			keyWithinSpace: 'Key within a space',
			functionWithinSpace: jest.fn(
				val => `Function within a space called with ${val}`
			),
		},
		UNCOMMON: {
			someSubspace: {
				keyWithinSubspace: 'ULTRA-RARE Key within a subspace',
				functionWithinSubspace: jest.fn(val => `SECRET-RARE ${val}`),
			},
			keyWithinSpace: 'RARE Key within a space',
			functionWithinSpace: jest.fn(val => `SUPER-RARE ${val}`),
		},
	},
};

describe('Translate', () => {
	let t = null;

	beforeAll(() => {
		t = translate({
			defaultSpace: 'COMMON',
			translations: testTranslations,
			language: Object.keys(testTranslations)[0],
		});
	});

	test('creates a translation function', () => {
		expect(t).toBeInstanceOf(Function);
	});

	test('provides translated text when supplied with matching arguments/keys', () => {
		expect(t('COMMON', 'someSubspace', 'keyWithinSubspace')).toMatchSnapshot(
			'COMMON.someSubspace.keyWithinSubspace'
		);
		expect(t('COMMON', null, 'keyWithinSpace')).toMatchSnapshot(
			'COMMON.keyWithinSpace'
		);
		expect(t('UNCOMMON', 'someSubspace', 'keyWithinSubspace')).toMatchSnapshot(
			'UNCOMMON.someSubspace.keyWithinSubspace'
		);
		expect(t('UNCOMMON', null, 'keyWithinSpace')).toMatchSnapshot(
			'UNCOMMON.keyWithinSpace'
		);
	});

	test('runs any functions and return values when they are called via the translation function', () => {
		const test1 = t(
			'COMMON',
			'someSubspace',
			'functionWithinSubspace'
		)('TEST-1');
		const test2 = t('COMMON', null, 'functionWithinSpace')('TEST-2');
		const test3 = t(
			'UNCOMMON',
			'someSubspace',
			'functionWithinSubspace'
		)('TEST-3');
		const test4 = t('UNCOMMON', null, 'functionWithinSpace')('TEST-4');

		expect(testTranslations.test.COMMON.functionWithinSpace).toHaveBeenCalled();
		expect(
			testTranslations.test.COMMON.someSubspace.functionWithinSubspace
		).toHaveBeenCalled();
		expect(
			testTranslations.test.UNCOMMON.functionWithinSpace
		).toHaveBeenCalled();
		expect(
			testTranslations.test.UNCOMMON.someSubspace.functionWithinSubspace
		).toHaveBeenCalled();

		expect(test1).toMatchSnapshot('Function within subspace, in default space');
		expect(test2).toMatchSnapshot('Function within default space');
		expect(test3).toMatchSnapshot(
			'Function within subspace, in non-default space'
		);
		expect(test4).toMatchSnapshot('Function within non-default space');
	});

	describe('Error handling', () => {
		test("should throw MissingI18nConfigError when translations aren't provided", () => {
			expect(() => translate({ language: 'test' })).toThrowError(
				MissingI18nConfigError
			);
		});

		test("should throw MissingI18nConfigError when language isn't provided", () => {
			expect(() => translate({ translations: {} })).toThrowError(
				MissingI18nConfigError
			);
		});

		test('should throw console error when [x][y][z] is missing from translations', () => {
			const restoreConsole = mockConsole();

			const result1 = t('SOMEWHAT-COMMON', 'someSubspace', 'keyWithinSubSpace');
			const result2 = t(
				'COMMON',
				'someNonExistentSubspace',
				'keyWithinSubSpace'
			);
			const result3 = t('COMMON', 'someSubspace', 'nonExistentKey');

			/* eslint-disable no-console */
			expect(console.error).toHaveBeenCalledTimes(2);
			expect(console.error).toHaveBeenNthCalledWith(1, expect.any(Error));
			expect(console.error).toHaveBeenNthCalledWith(2, expect.any(Error));
			/* eslint-enable no-console */

			expect(result1).toBeNull();
			expect(result2).toBeNull();
			expect(result3).toBeUndefined();

			restoreConsole();
		});
	});
});
