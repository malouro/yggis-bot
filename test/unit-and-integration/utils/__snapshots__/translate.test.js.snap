// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translate provides translated text when supplied with matching arguments/keys: COMMON.keyWithinSpace 1`] = `"Key within a space"`;

exports[`Translate provides translated text when supplied with matching arguments/keys: COMMON.someSubspace.keyWithinSubspace 1`] = `"Key within a subspace"`;

exports[`Translate provides translated text when supplied with matching arguments/keys: UNCOMMON.keyWithinSpace 1`] = `"RARE Key within a space"`;

exports[`Translate provides translated text when supplied with matching arguments/keys: UNCOMMON.someSubspace.keyWithinSubspace 1`] = `"ULTRA-RARE Key within a subspace"`;

exports[`Translate runs any functions and return values when they are called via the translation function: Function within default space 1`] = `"Function within a space called with TEST-2"`;

exports[`Translate runs any functions and return values when they are called via the translation function: Function within non-default space 1`] = `"SUPER-RARE TEST-4"`;

exports[`Translate runs any functions and return values when they are called via the translation function: Function within subspace, in default space 1`] = `"Function within a subspace called with TEST-1"`;

exports[`Translate runs any functions and return values when they are called via the translation function: Function within subspace, in non-default space 1`] = `"SECRET-RARE TEST-3"`;
